pkg/backend/snapshot.go

```diff
@@ -658,36 +709,12 @@ func NewSnapshotManager(persister SnapshotPersister, baseSnap *deploy.Snapshot)
 		done:             done,
 	}
 
-	go func() {
-		// True if we have elided writes since the last actual write.
-		hasElidedWrites := false
-
-		// Service each mutation request in turn.
-	serviceLoop:
-		for {
-			select {
-			case request := <-mutationRequests:
-				var err error
-				if request.mutator() {
-					err = manager.saveSnapshot()
-					hasElidedWrites = false
-				} else {
-					hasElidedWrites = true
-				}
-				request.result <- err
-			case <-cancel:
-				break serviceLoop
-			}
-		}
-
-		// If we still have elided writes once the channel has closed, flush the snapshot.
-		var err error
-		if hasElidedWrites {
-			logging.V(9).Infof("SnapshotManager: flushing elided writes...")
-			err = manager.saveSnapshot()
-		}
-		done <- err
-	}()
+	serviceLoop := manager.defaultServiceLoop
+	unsafeEnabled := os.Getenv(experimentalSnapshotManagerFlag) != ""
+	if unsafeEnabled {
+		serviceLoop = manager.unsafeServiceLoop
+	}
+	go serviceLoop(mutationRequests, done)
```

Comment:
@t0yv0 refactored to break out the methods into their own functions as suggested.